packages/tpetra/core/test/CrsMatrix/CrsMatrix_ReplaceDiagonal.cpp

```diff
@@ -0,0 +1,212 @@
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
+//
+// ************************************************************************
+// @HEADER
+
+#include <Teuchos_UnitTestHarness.hpp>
+#include <TpetraCore_ETIHelperMacros.h>
+
+#include <Tpetra_Core.hpp>
+#include <Tpetra_CrsMatrix.hpp>
+#include <Tpetra_Map.hpp>
+#include <Tpetra_Vector.hpp>
+
+#include <Tpetra_replaceDiagonalCrsMatrix.hpp>
+
+#include <MatrixMarket_Tpetra.hpp>
+
+namespace { // (anonymous)
+
+
+  // Unit test of replacing the diagonal of a Tpetra::CrsMatrix
+  TEUCHOS_UNIT_TEST_TEMPLATE_4_DECL( CrsMatrix, ReplaceDiagonal, Scalar, LO, GO, Node )
+  {
+    using Teuchos::Comm;
+    using Teuchos::outArg;
+    using Teuchos::RCP;
+    using Teuchos::rcp;
+    using Teuchos::REDUCE_MIN;
+    using Teuchos::REDUCE_SUM;
+    using Teuchos::reduceAll;
+    using std::cerr;
+    using std::endl;
+    using std::size_t;
+    typedef Tpetra::Map<LO, GO, Node> map_type;
+    typedef Tpetra::CrsMatrix<Scalar, LO, GO, Node> crs_matrix_type;
+    typedef Tpetra::Vector<Scalar, LO, GO, Node> vec_type;
+    typedef Tpetra::MatrixMarket::Writer<crs_matrix_type> writer_type;
+    typedef typename Teuchos::Array<LO>::size_type size_type;
+    typedef Teuchos::ScalarTraits<Scalar> STS;
+    typedef typename STS::magnitudeType MT;
+    typedef Teuchos::ScalarTraits<MT> STM;
+    const Scalar ZERO = STS::zero ();
+    const Scalar ONE = STS::one ();
+    typedef Teuchos::OrdinalTraits<LO> OT;
+    const LO INVALID = OT::invalid();
+    const bool extraDebug = false; // for extra-verbose output
+    int lclSuccess = 0;
+    int gblSuccess = 0;
+
+    RCP<const Comm<int> > comm = Tpetra::getDefaultComm ();
+    const int myRank = comm->getRank ();
+
+    if (myRank == 0) {
+      cerr << "Tpetra replaceDiagonal test" << endl
+           << "Create Map and matrix" << endl;
+    }
+
+    // do a FEM-type communication, then apply to a MultiVector containing the identity
+    // this will check non-trivial communication and test multivector apply
+    // get a comm
+    const size_t numImages = comm->getSize();
+    const size_t myImageID = comm->getRank();
+    if (numImages < 3) return;
```

Comment:
If you need exactly 3 processes, then make the CMakeLists.txt logic require 3 processes, and throw if you don't get the right number.  That will prevent spurious successes.